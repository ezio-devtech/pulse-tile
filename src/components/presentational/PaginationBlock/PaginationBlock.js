import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import _ from 'lodash/fp';

const PaginationBlock = ({ entriesPerPage, totalEntriesAmount, offset, setOffset }) => {
  const isFirstPage = offset < entriesPerPage;
  const isLastPage = offset >= totalEntriesAmount - entriesPerPage;
  const pagesAmount = Math.ceil(totalEntriesAmount / entriesPerPage);
  const currentPage = Math.floor(offset / entriesPerPage) + 1;

  const setPage = page => () => {
    if (page < 1) return setPage(1);
    if (page > pagesAmount) return setPage(pagesAmount);
    return setOffset((page - 1) * entriesPerPage)
  };

  const calculatePageNumber = (i, actualPage, paginationRange, totalPages) => {
    const halfWay = Math.ceil(paginationRange / 2);
    if (i === paginationRange) {
      return totalPages;
    } else if (i === 1) {
      return i;
    } else if (paginationRange < totalPages) {
      if (totalPages - halfWay < actualPage) {
        return (totalPages - paginationRange) + i;
      } else if (halfWay < actualPage) {
        return (actualPage - halfWay) + i;
      }
      return i;
    }
    return i;
  };

  const pagination = (actualPage, collectionLength, rowsPerPage) => {
    const pages = [];
    const maxSizePage = 6;
    const totalPages = Math.ceil(collectionLength / 10);
    let paginationRange;

    if (rowsPerPage > maxSizePage) {
      paginationRange = maxSizePage;
    } else {
      paginationRange = rowsPerPage;
    }

    const halfWay = Math.ceil(paginationRange / 2);
    let position;

    if (actualPage <= halfWay) {
      position = 'start';
    } else if (totalPages - halfWay < actualPage) {
      position = 'end';
    } else {
      position = 'middle';
    }

    const ellipsesNeeded = paginationRange < totalPages;
    let i = 1;
    while (i <= totalPages && i <= paginationRange) {
      const pageNumber = calculatePageNumber(i, actualPage, paginationRange, totalPages);

      const openingEllipsesNeeded = (i === 2 && (position === 'middle' || position === 'end'));
      const closingEllipsesNeeded = (i === paginationRange - 1 && (position === 'middle' || position === 'start'));
      if (ellipsesNeeded && (openingEllipsesNeeded || closingEllipsesNeeded)) {
        pages.push('...');
      } else {
        pages.push(pageNumber);
      }
      i += 1;
    }
    return pages;
  };

  return (<ul className="pagination-block">
    <li className={classNames('pagination-item arrow', { disabled: isFirstPage })}>
      <a className="pagination-link" onClick={isFirstPage ? _.noop : setPage(1)}>«</a>
    </li>
    <li className={classNames('pagination-item arrow short-show', { disabled: isFirstPage })}>
      <a className="pagination-link pp" onClick={isFirstPage ? _.noop : setPage(currentPage - 1)}>‹</a>
    </li>
    { pagination(currentPage, totalEntriesAmount, pagesAmount).map(pageIndex =>
      <li className={classNames('pagination-item short-show', { active: currentPage === pageIndex, disabled: pageIndex === '...' })} key={_.uniqueId('__PaginationBlock__li__')}>
        <a className="pagination-link" onClick={pageIndex === '...' ? _.noop : setPage(pageIndex)}>{pageIndex}</a>
      </li>
    )}
    <li className={classNames('pagination-item arrow short-show', { disabled: isLastPage })}>
      <a className="pagination-link nn" onClick={isLastPage ? _.noop : setPage(currentPage + 1)}>›</a>
    </li>
    <li className={classNames('pagination-item arrow', { disabled: isLastPage })}>
      <a className="pagination-link" onClick={isLastPage ? _.noop : setPage(pagesAmount)}>»</a>
    </li>
  </ul>)
};

PaginationBlock.propTypes = {
  entriesPerPage: PropTypes.number.isRequired,
  totalEntriesAmount: PropTypes.number.isRequired,
  offset: PropTypes.number.isRequired,
  setOffset: PropTypes.func.isRequired,
};

export default PaginationBlock
