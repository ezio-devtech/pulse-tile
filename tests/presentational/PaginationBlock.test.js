import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import PaginationBlock from '../../src/components/presentational/PaginationBlock/PaginationBlock';

Enzyme.configure({ adapter: new Adapter() });

const entriesPerPage = 10;
const totalEntriesAmount = 89;

const handleSetOffset = offset => console.log(offset);

describe('Component <PaginationBlock />', () => {
  it('should renders on the first page', () => {
    const offset = 0;
    const tree = renderer
      .create(
        <PaginationBlock
          offset={offset}
          entriesPerPage={entriesPerPage}
          totalEntriesAmount={totalEntriesAmount}
          setOffset={handleSetOffset}
        />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });


  it('should renders when we move to another page', () => {
    const offset = 80;
    const tree = renderer
      .create(
        <PaginationBlock
          offset={offset}
          entriesPerPage={entriesPerPage}
          totalEntriesAmount={totalEntriesAmount}
          setOffset={handleSetOffset}
        />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders on the last page', () => {
    const offset = 20;
    const tree = renderer
      .create(
        <PaginationBlock
          offset={offset}
          entriesPerPage={entriesPerPage}
          totalEntriesAmount={totalEntriesAmount}
          setOffset={handleSetOffset}
        />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Test PaginationBlock component with transitions', () => {
    let offset = 0;
    const setOffset = (page) => {
      offset = page;
    };
    const paginationBlock = shallow(
      <PaginationBlock
        offset={offset}
        entriesPerPage={entriesPerPage}
        totalEntriesAmount={totalEntriesAmount}
        setOffset={setOffset}
      />
    );
    expect(paginationBlock).toMatchSnapshot();
    paginationBlock.find('.pagination-link.nn').simulate('click');
    expect(paginationBlock).toMatchSnapshot();
  });


  it('Test PaginationBlock when we have less than six pages', () => {
    const offset = 10;
    const totalEntriesAmount = 40;
    const tree = renderer
      .create(
        <PaginationBlock
          offset={offset}
          entriesPerPage={entriesPerPage}
          totalEntriesAmount={totalEntriesAmount}
          setOffset={handleSetOffset}
        />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Test PaginationBlock when we are in the middle of the pagination block', () => {
    const offset = 30;
    const totalEntriesAmount = 120;
    const tree = renderer
      .create(
        <PaginationBlock
          offset={offset}
          entriesPerPage={entriesPerPage}
          totalEntriesAmount={totalEntriesAmount}
          setOffset={handleSetOffset}
        />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Test PaginationBlock when totalEntriesAmount less then offset', () => {
    const offset = 30;
    const totalEntriesAmount = 29;
    const tree = renderer
      .create(
        <PaginationBlock
          offset={offset}
          entriesPerPage={entriesPerPage}
          totalEntriesAmount={totalEntriesAmount}
          setOffset={handleSetOffset}
        />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

