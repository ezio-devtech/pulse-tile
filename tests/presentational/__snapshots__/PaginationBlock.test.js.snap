// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <PaginationBlock /> should be renders when totalEntriesAmount less then offset 1`] = `
<ul
  className="pagination-block"
>
  <li
    className="pagination-item arrow"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      «
    </a>
  </li>
  <li
    className="pagination-item arrow short-show"
  >
    <a
      className="pagination-link pp"
      onClick={[Function]}
    >
      ‹
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      1
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      2
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      3
    </a>
  </li>
  <li
    className="pagination-item arrow short-show disabled"
  >
    <a
      className="pagination-link nn"
      onClick={undefined}
    >
      ›
    </a>
  </li>
  <li
    className="pagination-item arrow disabled"
  >
    <a
      className="pagination-link"
      onClick={undefined}
    >
      »
    </a>
  </li>
</ul>
`;

exports[`Component <PaginationBlock /> should be renders when we are in the middle of the pagination block 1`] = `
<ul
  className="pagination-block"
>
  <li
    className="pagination-item arrow"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      «
    </a>
  </li>
  <li
    className="pagination-item arrow short-show"
  >
    <a
      className="pagination-link pp"
      onClick={[Function]}
    >
      ‹
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      1
    </a>
  </li>
  <li
    className="pagination-item short-show disabled"
  >
    <a
      className="pagination-link"
      onClick={undefined}
    >
      ...
    </a>
  </li>
  <li
    className="pagination-item short-show active"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      4
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      5
    </a>
  </li>
  <li
    className="pagination-item short-show disabled"
  >
    <a
      className="pagination-link"
      onClick={undefined}
    >
      ...
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      12
    </a>
  </li>
  <li
    className="pagination-item arrow short-show"
  >
    <a
      className="pagination-link nn"
      onClick={[Function]}
    >
      ›
    </a>
  </li>
  <li
    className="pagination-item arrow"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      »
    </a>
  </li>
</ul>
`;

exports[`Component <PaginationBlock /> should be renders when we have less than six pages 1`] = `
<ul
  className="pagination-block"
>
  <li
    className="pagination-item arrow"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      «
    </a>
  </li>
  <li
    className="pagination-item arrow short-show"
  >
    <a
      className="pagination-link pp"
      onClick={[Function]}
    >
      ‹
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      1
    </a>
  </li>
  <li
    className="pagination-item short-show active"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      2
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      3
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      4
    </a>
  </li>
  <li
    className="pagination-item arrow short-show"
  >
    <a
      className="pagination-link nn"
      onClick={[Function]}
    >
      ›
    </a>
  </li>
  <li
    className="pagination-item arrow"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      »
    </a>
  </li>
</ul>
`;

exports[`Component <PaginationBlock /> should be with transitions 1`] = `
<ul
  className="pagination-block"
>
  <li
    className="pagination-item arrow disabled"
  >
    <a
      className="pagination-link"
    >
      «
    </a>
  </li>
  <li
    className="pagination-item arrow short-show disabled"
  >
    <a
      className="pagination-link pp"
    >
      ‹
    </a>
  </li>
  <li
    className="pagination-item short-show active"
    key="__PaginationBlock__li__19"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      1
    </a>
  </li>
  <li
    className="pagination-item short-show"
    key="__PaginationBlock__li__20"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      2
    </a>
  </li>
  <li
    className="pagination-item short-show"
    key="__PaginationBlock__li__21"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      3
    </a>
  </li>
  <li
    className="pagination-item short-show"
    key="__PaginationBlock__li__22"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      4
    </a>
  </li>
  <li
    className="pagination-item short-show disabled"
    key="__PaginationBlock__li__23"
  >
    <a
      className="pagination-link"
    >
      ...
    </a>
  </li>
  <li
    className="pagination-item short-show"
    key="__PaginationBlock__li__24"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      9
    </a>
  </li>
  <li
    className="pagination-item arrow short-show"
  >
    <a
      className="pagination-link nn"
      onClick={[Function]}
    >
      ›
    </a>
  </li>
  <li
    className="pagination-item arrow"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      »
    </a>
  </li>
</ul>
`;

exports[`Component <PaginationBlock /> should be with transitions 2`] = `
<ul
  className="pagination-block"
>
  <li
    className="pagination-item arrow disabled"
  >
    <a
      className="pagination-link"
    >
      «
    </a>
  </li>
  <li
    className="pagination-item arrow short-show disabled"
  >
    <a
      className="pagination-link pp"
    >
      ‹
    </a>
  </li>
  <li
    className="pagination-item short-show active"
    key="__PaginationBlock__li__19"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      1
    </a>
  </li>
  <li
    className="pagination-item short-show"
    key="__PaginationBlock__li__20"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      2
    </a>
  </li>
  <li
    className="pagination-item short-show"
    key="__PaginationBlock__li__21"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      3
    </a>
  </li>
  <li
    className="pagination-item short-show"
    key="__PaginationBlock__li__22"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      4
    </a>
  </li>
  <li
    className="pagination-item short-show disabled"
    key="__PaginationBlock__li__23"
  >
    <a
      className="pagination-link"
    >
      ...
    </a>
  </li>
  <li
    className="pagination-item short-show"
    key="__PaginationBlock__li__24"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      9
    </a>
  </li>
  <li
    className="pagination-item arrow short-show"
  >
    <a
      className="pagination-link nn"
      onClick={[Function]}
    >
      ›
    </a>
  </li>
  <li
    className="pagination-item arrow"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      »
    </a>
  </li>
</ul>
`;

exports[`Component <PaginationBlock /> should renders on the first page 1`] = `
<ul
  className="pagination-block"
>
  <li
    className="pagination-item arrow disabled"
  >
    <a
      className="pagination-link"
      onClick={undefined}
    >
      «
    </a>
  </li>
  <li
    className="pagination-item arrow short-show disabled"
  >
    <a
      className="pagination-link pp"
      onClick={undefined}
    >
      ‹
    </a>
  </li>
  <li
    className="pagination-item short-show active"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      1
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      2
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      3
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      4
    </a>
  </li>
  <li
    className="pagination-item short-show disabled"
  >
    <a
      className="pagination-link"
      onClick={undefined}
    >
      ...
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      9
    </a>
  </li>
  <li
    className="pagination-item arrow short-show"
  >
    <a
      className="pagination-link nn"
      onClick={[Function]}
    >
      ›
    </a>
  </li>
  <li
    className="pagination-item arrow"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      »
    </a>
  </li>
</ul>
`;

exports[`Component <PaginationBlock /> should renders on the last page 1`] = `
<ul
  className="pagination-block"
>
  <li
    className="pagination-item arrow"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      «
    </a>
  </li>
  <li
    className="pagination-item arrow short-show"
  >
    <a
      className="pagination-link pp"
      onClick={[Function]}
    >
      ‹
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      1
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      2
    </a>
  </li>
  <li
    className="pagination-item short-show active"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      3
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      4
    </a>
  </li>
  <li
    className="pagination-item short-show disabled"
  >
    <a
      className="pagination-link"
      onClick={undefined}
    >
      ...
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      9
    </a>
  </li>
  <li
    className="pagination-item arrow short-show"
  >
    <a
      className="pagination-link nn"
      onClick={[Function]}
    >
      ›
    </a>
  </li>
  <li
    className="pagination-item arrow"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      »
    </a>
  </li>
</ul>
`;

exports[`Component <PaginationBlock /> should renders when we move to another page 1`] = `
<ul
  className="pagination-block"
>
  <li
    className="pagination-item arrow"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      «
    </a>
  </li>
  <li
    className="pagination-item arrow short-show"
  >
    <a
      className="pagination-link pp"
      onClick={[Function]}
    >
      ‹
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      1
    </a>
  </li>
  <li
    className="pagination-item short-show disabled"
  >
    <a
      className="pagination-link"
      onClick={undefined}
    >
      ...
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      6
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      7
    </a>
  </li>
  <li
    className="pagination-item short-show"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      8
    </a>
  </li>
  <li
    className="pagination-item short-show active"
  >
    <a
      className="pagination-link"
      onClick={[Function]}
    >
      9
    </a>
  </li>
  <li
    className="pagination-item arrow short-show disabled"
  >
    <a
      className="pagination-link nn"
      onClick={undefined}
    >
      ›
    </a>
  </li>
  <li
    className="pagination-item arrow disabled"
  >
    <a
      className="pagination-link"
      onClick={undefined}
    >
      »
    </a>
  </li>
</ul>
`;
